tests/test_logicals.py

```diff
@@ -86,6 +90,46 @@ def test_decimal():
     assert_schema(py_type, expected)
 
 
+def test_annotated_decimal():
+    py_type = Annotated[decimal.Decimal, (5, 2)]
+    expected = {
+        "type": "bytes",
+        "logicalType": "decimal",
+        "precision": 5,
+        "scale": 2,
+    }
+    assert_schema(py_type, expected)
+
+
+def test_annotated_decimal_neg_scale():
+    py_type = Annotated[decimal.Decimal, (5, -2)]
+    expected = {
+        "type": "bytes",
+        "logicalType": "decimal",
+        "precision": 5,
+        "scale": -2,
+    }
+    assert_schema(py_type, expected)
+
+
+def test_annotated_decimal_no_tuple():
+    py_type = Annotated[decimal.Decimal, ...]
+    with pytest.raises(pas.TypeNotSupportedError):
+        assert_schema(py_type, {})
+
+
+def test_annotated_decimal_tuple_wrong_length():
+    py_type = Annotated[decimal.Decimal, (3, 2, 1)]
+    with pytest.raises(pas.TypeNotSupportedError):
+        assert_schema(py_type, {})
+
+
+def test_annotated_decimal_tuple_wrong_type():
+    py_type = Annotated[decimal.Decimal, ("a", 1)]
+    with pytest.raises(pas.TypeNotSupportedError):
+        assert_schema(py_type, {})
```

Comment:
Would it make sense to have tests with all supported data classes to validate the type works?

So one for dataclasses and one for pydantic?

Reply:
that would only be necessary for the default value. The type itself can be tested without a data structure around it.

Reply:
hmm looks like it doesn't work like I though it would. this throws an error

```
from decimal import Decimal
from typing import Annotated

import py_avro_schema as pas
from pydantic import BeforeValidator, BaseModel


class Foo(BaseModel):
    bar: Annotated[Decimal, (3, 2), BeforeValidator(lambda x: x)]


if __name__ == "__main__":
    print(Foo(bar=Decimal("1.23")))
    print(pas.generate(Foo))
```

Pydantic works, but pas.generate throws

```
py_avro_schema._schemas.TypeNotSupportedError: Cannot generate Avro schema for Python type <class 'decimal.Decimal'>
```

What am I doing wrong?

Reply:
@faph this still throws an error with the new DecimalMeta class used. Looks like py_type is decimal, and origin is none: what do i do wrong? 🤔

Reply:
It's because of this PydanticSchema code

```
# Pydantic 2 resolves forward references for us. To avoid infinite recursion, we check if the unresolved raw
        # annoation is a forward reference. If so, we use that instead of Pydantic's resolved type hint. There might be
        # a better way to un-resolve the forward reference...
        # This does not support forward references from base model classes. If required we might need to traverse class
        # hierarchy to get the raw annotations?
        if isinstance(self.raw_annotations.get(name), (str, ForwardRef)):
            py_type = self.raw_annotations[name]
        else:
            py_type = py_field.annotation
```

py_field.annotation is only decimal. The raw_annotations contains the fully annotated type definition.

Reply:
Thanks for debugging this! Let me add a test to cover this.

I wonder whether we should be using raw annotations with Pydantic in general...

Reply:
This is actually very useful as this forces us to handle Annotated types generically. I have now changed this to use "raw" annotations for all Pydantic fields.

Reply:
I have raised a separate issue #60 to replace all subclass checks with the better check to include annotated types, throughout.

Reply:
@dada-engineer Just to say, I have slightly re-jigged the logic again in #62. This change includes that **any** `decimal.Decimal` is not caught by `handles_type` for decimal schema. That means you will get a better error message if you define a Python decimal type without the correct meta data, instead of raising a type-not-supported error.

Reply:
In any case, let me know if v `3.4.0` works ok for you @dada-engineer

Reply:
@faph @jcameron73  currently with pydantic it is not usable in the scenario of combining decimal and other datatypes in a union type like Decimal | None:

If I do the following I still get can not generate avro schema for type Decimal

```
OptionalAvroDecimal = Annotated[
    Decimal | None,
    pas.DecimalMeta(precision=DEFAULT_PRECISION, scale=DEFAULT_SCALE),
    BeforeValidator(
        lambda x: some_validator(x)
    ),
]
```

Which is okay I guess because what I really want is this:

```
OptionalAvroDecimal = Annotated[
    Decimal,
    pas.DecimalMeta(precision=DEFAULT_PRECISION, scale=DEFAULT_SCALE),
    BeforeValidator(
        lambda x: some_validator(x)
    ),
] | None
```

i.e. defining the validators for the decimal only and annotate decimal only with an option to be None. But this can not be done currently as it will raise the following error:

```
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:1368: in __or__
    return Union[self, right]
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:355: in inner
    return func(*args, **kwds)
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:478: in __getitem__
    return self._getitem(self, parameters)
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:690: in Union
    parameters = _remove_dups_flatten(parameters)
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:324: in _remove_dups_flatten
    return tuple(_deduplicate(params))
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:299: in _deduplicate
    all_params = set(params)
../../../.pyenv/versions/3.11.4/lib/python3.11/typing.py:2141: in __hash__
    return hash((self.__origin__, self.__metadata__))
E   TypeError: unhashable type: 'DecimalMeta'
```

This is currently the dealbreaker for me to say it is working.

Reply:
Sounds like decimal meta should just be a frozen/immutable type?

Do you have a unit test to reproduce ^^^?

Reply:
No unfortunately not. Shouldn't be to hard though 😁.

Reply:
Let me try :)

Reply:
Got it. So this is just because it's used in a union. And unioned types must be hashable it seems...

Reply:
Fixed here @dada-engineer : https://github.com/jpmorganchase/py-avro-schema/pull/62/commits/f8ced267fdd22c57ddc7a7c1d450442842acfa7d

Reply:
I am wondering whether this hashable requirement is not documented here: https://docs.python.org/3/library/typing.html?highlight=typing#typing.Annotated

Reply:
works for my Usecase @faph

Reply:
Perfect

Reply:
Raised the documentation issue here: https://github.com/python/cpython/issues/112281